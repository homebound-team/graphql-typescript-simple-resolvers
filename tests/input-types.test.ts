import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("Input Types & Arguments", () => {
  const schemaWithInputs = `
    scalar Date
    scalar DateTime

    enum Status {
      ACTIVE
      INACTIVE
      PENDING
    }

    input AddressInput {
      street: String!
      city: String!
      state: String
      zipCode: String
      country: String!
    }

    input UserInput {
      name: String!
      email: String!
      age: Int
      birthday: Date
      status: Status
      isActive: Boolean
      address: AddressInput
      tags: [String!]
      optionalTags: [String]
      permissions: [String!]!
      metadata: [String]!
    }

    input UpdateUserInput {
      name: String
      email: String
      age: Int
      status: Status
    }

    input SearchFilters {
      minAge: Int
      maxAge: Int
      statuses: [Status!]
      isActive: Boolean
    }

    type User {
      id: ID!
      name: String!
      email: String!
      age: Int
      birthday: Date
      status: Status
    }

    type Query {
      # Scalar arguments
      user(id: ID!): User
      userByEmail(email: String!): User

      # Optional arguments
      users(limit: Int, offset: Int): [User!]!

      # Input type arguments
      searchUsers(filters: SearchFilters): [User!]!
      getUsersInRange(minAge: Int!, maxAge: Int!): [User!]!

      # Array arguments
      usersByIds(ids: [ID!]!): [User!]!
      usersByOptionalIds(ids: [ID]): [User!]!
    }

    type Mutation {
      createUser(input: UserInput!): User!
      updateUser(id: ID!, input: UpdateUserInput!): User!

      # Multiple arguments including inputs
      createUserWithAddress(
        userInput: UserInput!
        addressInput: AddressInput!
        sendWelcomeEmail: Boolean
      ): User!
    }
  `;

  it("generates input types with proper nullability", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: { Date: "Date", DateTime: "Date" },
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface UserInput {
  name: string;
  email: string;
  age?: number | null | undefined;
  birthday?: Date | null | undefined;
  status?: Status | null | undefined;
  isActive?: boolean | null | undefined;
  address?: AddressInput | null | undefined;
  tags?: string[] | null | undefined;
  optionalTags?: Array<string | null | undefined> | null | undefined;
  permissions: string[];
  metadata: Array<string | null | undefined>;
}`);
  });

  it("generates nested input types", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface AddressInput {
  street: string;
  city: string;
  state?: string | null | undefined;
  zipCode?: string | null | undefined;
  country: string;
}`);
  });

  it("generates query resolvers with scalar arguments", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`user: Resolver<{}, QueryUserArgs, User | null | undefined>;`);
    expect(code).toContain(`userByEmail: Resolver<{}, QueryUserByEmailArgs, User | null | undefined>;`);
  });

  it("generates query resolvers with optional arguments", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`users: Resolver<{}, QueryUsersArgs, readonly User[]>;`);
  });

  it("generates query resolvers with array arguments", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`usersByIds: Resolver<{}, QueryUsersByIdsArgs, readonly User[]>;`);
    expect(code).toContain(`usersByOptionalIds: Resolver<{}, QueryUsersByOptionalIdsArgs, readonly User[]>;`);
  });

  it("generates mutation resolvers with input arguments", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`createUser: Resolver<{}, MutationCreateUserArgs, User>;`);
    expect(code).toContain(`updateUser: Resolver<{}, MutationUpdateUserArgs, User>;`);
  });

  it("generates argument type definitions", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface QueryUserArgs {
  id: string;
}`);

    expect(code).toContain(`export interface QueryUsersArgs {
  limit?: number | null | undefined;
  offset?: number | null | undefined;
}`);

    expect(code).toContain(`export interface MutationCreateUserArgs {
  input: UserInput;
}`);
  });

  it("snapshots input type generation", async () => {
    const schema = createTestSchema(schemaWithInputs);
    const code = await runPlugin(schema, {
      scalars: { Date: "Date", DateTime: "Date" },
      mappers: {},
      enumValues: {},
    });
    expect(code).toMatchSnapshot();
  });
});
