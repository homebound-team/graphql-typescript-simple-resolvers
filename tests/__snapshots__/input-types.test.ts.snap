// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Input Types & Arguments snapshots input type generation 1`] = `
"import { GraphQLResolveInfo, GraphQLScalarType } from "graphql";
import { Context } from "./context";

export interface Resolvers {
  Query: QueryResolvers;
  Mutation: MutationResolvers;
  User?: UserResolvers;
  Date: GraphQLScalarType;
  DateTime: GraphQLScalarType;
}

export type UnionResolvers = {};

export interface QueryResolvers {
  user: Resolver<{}, QueryUserArgs, User | null | undefined>;
  userByEmail: Resolver<{}, QueryUserByEmailArgs, User | null | undefined>;
  users: Resolver<{}, QueryUsersArgs, readonly User[]>;
  searchUsers: Resolver<{}, QuerySearchUsersArgs, readonly User[]>;
  getUsersInRange: Resolver<{}, QueryGetUsersInRangeArgs, readonly User[]>;
  usersByIds: Resolver<{}, QueryUsersByIdsArgs, readonly User[]>;
  usersByOptionalIds: Resolver<{}, QueryUsersByOptionalIdsArgs, readonly User[]>;
}

export interface MutationResolvers {
  createUser: Resolver<{}, MutationCreateUserArgs, User>;
  updateUser: Resolver<{}, MutationUpdateUserArgs, User>;
  createUserWithAddress: Resolver<{}, MutationCreateUserWithAddressArgs, User>;
}

export interface UserResolvers {
  id: Resolver<User, {}, string>;
  name: Resolver<User, {}, string>;
  email: Resolver<User, {}, string>;
  age: Resolver<User, {}, number | null | undefined>;
  birthday: Resolver<User, {}, Date | null | undefined>;
  status: Resolver<User, {}, Status | null | undefined>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface QueryUserArgs {
  id: string;
}
export interface QueryUserByEmailArgs {
  email: string;
}
export interface QueryUsersArgs {
  limit?: number | null | undefined;
  offset?: number | null | undefined;
}
export interface QuerySearchUsersArgs {
  filters?: SearchFilters | null | undefined;
}
export interface QueryGetUsersInRangeArgs {
  minAge: number;
  maxAge: number;
}
export interface QueryUsersByIdsArgs {
  ids: readonly string[];
}
export interface QueryUsersByOptionalIdsArgs {
  ids?: ReadonlyArray<string | null | undefined> | null | undefined;
}
export interface MutationCreateUserArgs {
  input: UserInput;
}
export interface MutationUpdateUserArgs {
  id: string;
  input: UpdateUserInput;
}
export interface MutationCreateUserWithAddressArgs {
  userInput: UserInput;
  addressInput: AddressInput;
  sendWelcomeEmail?: boolean | null | undefined;
}
export interface User {
  id: string;
  name: string;
  email: string;
  age: number | null | undefined;
  birthday: Date | null | undefined;
  status: Status | null | undefined;
}

export interface AddressInput {
  street: string;
  city: string;
  state?: string | null | undefined;
  zipCode?: string | null | undefined;
  country: string;
}

export interface UserInput {
  name: string;
  email: string;
  age?: number | null | undefined;
  birthday?: Date | null | undefined;
  status?: Status | null | undefined;
  isActive?: boolean | null | undefined;
  address?: AddressInput | null | undefined;
  tags?: string[] | null | undefined;
  optionalTags?: Array<string | null | undefined> | null | undefined;
  permissions: string[];
  metadata: Array<string | null | undefined>;
}

export interface UpdateUserInput {
  name?: string | null | undefined;
  email?: string | null | undefined;
  age?: number | null | undefined;
  status?: Status | null | undefined;
}

export interface SearchFilters {
  minAge?: number | null | undefined;
  maxAge?: number | null | undefined;
  statuses?: Status[] | null | undefined;
  isActive?: boolean | null | undefined;
}

export enum Status {
  Active = "ACTIVE",
  Inactive = "INACTIVE",
  Pending = "PENDING",
}
"
`;
